### PR TITLE
Added option to Clean the project in a Xamarin Android Build Task.

### PR DESCRIPTION
**Pull request by azchohfi**
It was already possible to do this, using "Clean" as target, but it is much easier from a UX perspective to use a checkbox, just like MSBuild Task does. All localized strings were copied from MSBuild Task, so they should be correct.

### CODE REVIEW

File: Tasks/XamarinAndroid/Strings/resources.resjson/de-de/resources.resjson [hunk 1]
```diff
@@ -11,6 +11,8 @@
   "loc.input.help.target": "Erstellt diese Ziele in diesem Projekt. Trennen Sie mehrere Ziele mit einem Semikolon.",
   "loc.input.label.outputDir": "Ausgabeverzeichnis",
   "loc.input.label.configuration": "Konfiguration",
+  "loc.input.label.clean": "Bereinigen",
+  "loc.input.help.clean": "Führt einen bereinigten Build (\"/t:clean\") vor dem Build aus.",
```
Comment (madhurig) [inline Tasks/XamarinAndroid/Strings/resources.resjson/de-de/resources.resjson]: Did you come up with the localized strings? You shouldn't need to do that, only update the task.json with ENU strings and rest will be generated later on.

File: Tasks/XamarinAndroid/Strings/resources.resjson/es-es/resources.resjson [hunk 2]
```diff
@@ -11,6 +11,8 @@
   "loc.input.help.target": "Compile estos destinos en este proyecto. Use puntos y comas para separar distintos destinos.",
   "loc.input.label.outputDir": "Directorio de salida",
   "loc.input.label.configuration": "Configuración",
+  "loc.input.label.clean": "Limpiar",
+  "loc.input.help.clean": "Ejecute una compilación limpia (/t:clean) antes de realizar la compilación.",
```
Comment (madhurig) [inline Tasks/XamarinAndroid/Strings/resources.resjson/es-es/resources.resjson]: Non ENU strings should not be updated as mentioned above

File: Tasks/XamarinAndroid/Strings/resources.resjson/fr-fr/resources.resjson [hunk 3]
```diff
@@ -11,6 +11,8 @@
   "loc.input.help.target": "Générez ces cibles dans ce projet. Utilisez un point-virgule pour séparer plusieurs cibles.",
   "loc.input.label.outputDir": "Répertoire de sortie",
   "loc.input.label.configuration": "Configuration",
+  "loc.input.label.clean": "Nettoyer",
```
Comment (madhurig) [inline Tasks/XamarinAndroid/Strings/resources.resjson/fr-fr/resources.resjson]: Non ENU strings should not be updated as mentioned above

File: Tasks/XamarinAndroid/Strings/resources.resjson/it-IT/resources.resjson [hunk 4]
```diff
@@ -11,6 +11,8 @@
   "loc.input.help.target": "Consente di compilare queste destinazioni nel progetto. Usare il punto e virgola per separare più destinazioni.",
   "loc.input.label.outputDir": "Directory di output",
   "loc.input.label.configuration": "Configurazione",
+  "loc.input.label.clean": "Pulisci",
+  "loc.input.help.clean": "Esegue una compilazione pulita (/t:clean) prima della compilazione.",
```
Comment (madhurig) [inline Tasks/XamarinAndroid/Strings/resources.resjson/it-IT/resources.resjson]: Non ENU strings should not be updated as mentioned above

File: Tasks/XamarinAndroid/Strings/resources.resjson/ja-jp/resources.resjson [hunk 5]
```diff
@@ -11,6 +11,8 @@
   "loc.input.help.target": "指定したターゲットをこのプロジェクトでビルドします。ターゲットが複数の場合には、セミコロンを使用して区切ります。",
   "loc.input.label.outputDir": "出力ディレクトリ",
   "loc.input.label.configuration": "構成",
+  "loc.input.label.clean": "消去",
```
Comment (madhurig) [inline Tasks/XamarinAndroid/Strings/resources.resjson/ja-jp/resources.resjson]: Non ENU strings should not be updated as mentioned above

File: Tasks/XamarinAndroid/Strings/resources.resjson/ko-KR/resources.resjson [hunk 6]
```diff
@@ -11,6 +11,8 @@
   "loc.input.help.target": "이 프로젝트에서 이 대상을 빌드하세요. 여러 대상을 구분하려면 세미콜론을 사용하세요.",
   "loc.input.label.outputDir": "출력 디렉터리",
   "loc.input.label.configuration": "구성",
+  "loc.input.label.clean": "클린",
```
Comment (madhurig) [inline Tasks/XamarinAndroid/Strings/resources.resjson/ko-KR/resources.resjson]: Non ENU strings should not be updated as mentioned above

File: Tasks/XamarinAndroid/Strings/resources.resjson/ru-RU/resources.resjson [hunk 7]
```diff
@@ -11,6 +11,8 @@
   "loc.input.help.target": "Выполнить сборку этих целевых объектов в этом проекте. Несколько целевых объектов указываются через точку с запятой.",
   "loc.input.label.outputDir": "Выходной каталог",
   "loc.input.label.configuration": "Конфигурация",
+  "loc.input.label.clean": "Очистить",
+  "loc.input.help.clean": "Выполните очистку сборки (/t:clean) перед сборкой.",
```
Comment (madhurig) [inline Tasks/XamarinAndroid/Strings/resources.resjson/ru-RU/resources.resjson]: Non ENU strings should not be updated as mentioned above

File: Tasks/XamarinAndroid/Strings/resources.resjson/zh-CN/resources.resjson [hunk 8]
```diff
@@ -11,6 +11,8 @@
   "loc.input.help.target": "在此项目中生成这些目标。使用分号分隔多个目标。",
   "loc.input.label.outputDir": "输出目录",
   "loc.input.label.configuration": "配置",
+  "loc.input.label.clean": "清理",
```
Comment (madhurig) [inline Tasks/XamarinAndroid/Strings/resources.resjson/zh-CN/resources.resjson]: Non ENU strings should not be updated as mentioned above

File: Tasks/XamarinAndroid/Strings/resources.resjson/zh-TW/resources.resjson [hunk 9]
```diff
@@ -11,6 +11,8 @@
   "loc.input.help.target": "在此專案中建置這些目標。請使用分號以區隔多個目標。",
   "loc.input.label.outputDir": "輸出目錄",
   "loc.input.label.configuration": "組態",
+  "loc.input.label.clean": "清除",
+  "loc.input.help.clean": "在組建之前執行乾淨的組建 (/t:clean)。",
```
Comment (madhurig) [inline Tasks/XamarinAndroid/Strings/resources.resjson/zh-TW/resources.resjson]: Non ENU strings should not be updated as mentioned above

**Comment by madhurig** [comment]
@azchohfi : Thank you for suggesting an improvement and making the changes to the task. Overall looks good to me except the resource strings. Can you please attach a screenshot of the task with the "Clean" input.

**Comment by azchohfi** [comment]
Hello @madhurig! Thanks for the feedback. I've used the strings from MSBuild Task (https://github.com/Microsoft/vsts-tasks/tree/master/Tasks/MSBuild/Strings/resources.resjson). I've assumed they are correct since it's the exact same functionality. Should I remove them anyway?

Screenshot: 
image

**Comment by madhurig** [comment]
@azchohfi : thanks for sharing the screenshot, it looks good. Yes, please remove the non ENU strings that you added since it might interfere with the localization process on the repo. 

Thanks,
Madhuri

**Comment by azchohfi** [comment]
Done.

**Comment by madhurig** [comment]
@azchohfi : I double checked the implementation yesterday for msbuild. When "Clean" is set to true in msbuild, the task first runs msbuild /t:Clean and then runs msbuild with whatever targets the user specified. To keep this behavior consistent with msbuild task and the description of the "Clean" option, you will need to do the same. First run Xamarin Android with /t:Clean only and then run the build with the /t:PackageForAndroid. More testing might be needed to identify if there is any difference in output.

Thanks,
Madhuri

**Comment by azchohfi** [comment]
Just to make sure that we are on the same page, I've also found this Invoke-BuildTools implementation, and yes, it does invoke MSBuild two times if clean is enabled.

The only reason, in my limited knowledge, for this behavior, is to log the cleaning inside a different file (https://github.com/Microsoft/vsts-tasks/blob/b5b5712b818c5dd93eed63558d2860238e4aba77/Tasks/Common/MSBuildHelpers/InvokeFunctions.ps1#L32). Should I just be careful about this, or is there any other reason for this behavior? I ask this because MSBuild does support multiple targets and the *order matters*, so if logging is not important, then I should only be careful about the order of the Clean target (which I did in fact made a mistake on the order of the TypeScript implementation (https://github.com/azchohfi/vsts-tasks/blob/0920097e18cef8f7915c847f8c90abb78f719622/Tasks/XamarinAndroid/xamarinandroid.ts#L71).

**Comment by azchohfi** [comment]
BTW, we should start this work on iOS build as well, but let me finish this one until I can start something else.

**Comment by azchohfi** [comment]
I changed the order of /tClean to make sure the behavior is the same on PowerShell as it is on TypeScript.

**Comment by azchohfi** [comment]
@madhurig any insight?

File: Tasks/XamarinAndroid/task.json [hunk 10]
```diff
@@ -65,6 +65,14 @@
             "required": false
         },
         {
+            "name": "clean",
+            "type": "boolean",
+            "label": "Clean",
+            "defaultValue": "false",
+            "required": false,
+            "helpMarkDown": "Run a clean build (/t:clean) prior to the build."
```
Comment (madhurig) [inline Tasks/XamarinAndroid/task.json]: Change this to "Passes the clean target (/t:clean) during build."

**Comment by madhurig** [comment]
@azchohfi: I feel it is fine to just run the build with /t:Clean and then any other targets. I would update the help text to reflect this as suggested above. Thanks for verifying and taking the time to make these changes. 

-Madhuri

**Comment by azchohfi** [comment]
Done!

**Comment by madhurig** [comment]
Looks good, please run a clean build to update the string in task.loc.json and we should be good. Thanks for all your work on this.

**Comment by azchohfi** [comment]
Done!

**Comment by madhurig** [comment]
@azchohfi: Thanks for your contribution!

**Pull request merged**